reps/2022-12-15-optimizer-data.md

```diff
@@ -0,0 +1,308 @@
+# Execution Optimizer for Ray Datasets
+
+## Summary
+
+Build the breakthrough foundation to tackle a series of fundamental issues around Ray Data. The foundation is (1) **lazy execution**, (2) **optimizer**, and (3) **vectorized execution with data batch**.
+
+### General Motivation
+Two key ML workloads for Datasets are (1) **data ingest**, where trainer processes (e.g., PyTorch workers), read, preprocess, and ingest data, and (2) **batch inference**, where we want to apply a pretrained model across a large dataset to generate predictions.
+
+Currently Ray Data does not work out-of-box for both workloads with followed reasons:
+* Data batch size and parallelism is challenging to tune by users, often resulting in poor performance and frustration.
+* Expensive data conversion, copy, and materialization to object store, can happen everywhere.
+* Lack of optimizer for execution plan, caused too much hand-holding with users to write the best program by hand.
+
+To resolve the above issues, we need to change Ray Datasets in multiple perspectives:
+* User experience:
+  * Configurations auto-tuning: auto-tuning critical configurations such as data batch size and parallelism. Fewer configurations and less confusing for users.
+  * Minimize cognitive load: lazy-execution first and only.
+* Performance
+  * Memory efficiency: zero-copy batch iterator. Operators pushdown. Eliminate unnecessary batch copy and format conversion.
+* Extensibility
+  * Easier to develop other optimization in the future
+    * Special optimizer for AIR workload.
+    * Example: prune unused columns through the execution plan as early as possible (as ColumnPruningPushDownRule).
+* Simplify code base
+  * Unify `LazyBlockList` and `BlockList`, delete `StatsActor`, eager mode and other legacy behavior.
+  * Able to support more Datasets APIs.
+
+### Should this change be within `ray` or outside?
+
+The proposed change is inside the code of Ray Datasets, which is already part of the Ray project repository (“ray.data”).
+
+## Stewardship
+### Required Reviewers
+
+@clarkzinzow, @jianoaix, @ericl, @stephanie-wang
+
+### Shepherd of the Proposal (should be a senior committer)
+
+@ericl
+
+## Design and Architecture
+
+Architecture after REP:
+
+<img width="945" alt="new-architecture" src="https://user-images.githubusercontent.com/4629931/207807703-bb65db63-41a0-41d9-8e7b-154e1a0ed565.png">
```

Comment:
Since we are considering re-optimization outside the scope of this REP, can we also remove that from the diagram?

Reply:
yeah, removed.